libqtile/widget/decorations.py

```diff
@@ -0,0 +1,226 @@
+
+        degrees = math.pi / 180.0
+
+        self.ctx.new_sub_path()
+
+        delta = self.radius + self.linewidth / 2
+
+        self.ctx.arc(
+            self.padding_x + box_width - delta,
+            self.padding_y + delta,
+            self.radius,
+            -90 * degrees,
+            0 * degrees
+        )
+
+        self.ctx.arc(
+            self.padding_x + box_width - delta,
+            self.padding_y + box_height - delta,
+            self.radius,
+            0 * degrees,
+            90 * degrees
+        )
+
+        self.ctx.arc(
+            self.padding_x + delta,
+            self.padding_y + box_height - delta,
+            self.radius,
+            90 * degrees,
+            180 * degrees
+        )
+        self.ctx.arc(
+            self.padding_x + delta,
+            self.padding_y + delta,
+            self.radius,
+            180 * degrees,
+            270 * degrees
+        )
+
+        self.ctx.close_path()
+
+        if self.filled:
+            self.ctx.fill()
+        else:
+            self.ctx.set_line_width(self.linewidth)
+            self.ctx.stroke()
+
+
+class BorderDecoration(_Decoration):
+    """
+        Widget decoration that draws a straight line on the widget border.
+        Padding can be used to adjust the position of the border further.
+
+        Only one colour can be set but decorations can be layered to achieve
+        multi-coloured effects.
+    """
+    defaults = [
+        ("colour", "#000000", "Border colour"),
+        (
+            "border_width",
+            2,
+            "Border width. Single number is all edges, or [top/bottom, left/right] or [top, bottom, left, right]"
+        )
+    ]
+
+    def __init__(self, **config):
+        _Decoration.__init__(self, **config)
+        self.add_defaults(BorderDecoration.defaults)
+
+        if type(self.border_width) in [float, int]:
+            tp = bm = lt = rt = self.border_width
+        elif type(self.border_width) in [tuple, list]:
+            if len(self.border_width) == 1:
+                tp = bm = lt = rt = self.border_width[0]
+            elif len(self.border_width) == 2:
+                tp = bm = self.border_width[0]
+                lt = rt = self.border_width[1]
+            elif len(self.border_width) == 4:
+                tp, bm, lt, rt = self.border_width
+            else:
+                logger.info("Border width should be a single number or a list of 1, 2 or 4 values")
+                tp = bm = lt = rt = 0
+        else:
+            logger.info("Border width should be a single number or a list of 1, 2 or 4 values")
+            tp = bm = lt = rt = 0
+
+        self.borders = [tp, bm, lt, rt]
+
+    def draw(self, parent):
+        _Decoration.draw(self, parent)
+
+        top, bottom, left, right = self.borders
+
+        self.drawer.set_source_rgb(self.colour)
+
+        if top:
+            offset = top // 2
+            self._draw_border(
+                offset + self.padding_x,
```

Comment:
Think my maths is right but still getting slight gaps between widgets.

Reply:
Could it be due to the `//` operators and a rounding error?

Reply:
Yeah, tried that. Will do some more digging.